tools/wptrunner/wptrunner/executors/executorcontentshell.py

```diff
@@ -0,0 +1,211 @@
+# mypy: allow-untyped-defs
+
+from .base import RefTestExecutor, RefTestImplementation, CrashtestExecutor, TestharnessExecutor
+from .protocol import ConnectionlessProtocol
+from time import time
+from queue import Empty
+from base64 import b64encode
+from os import linesep
+import json
+
+class CrashError(BaseException):
+    pass
+
+class ContentShellProtocol(ConnectionlessProtocol):
+    """This class represents the protocol used by content_shell in protocol mode.
+    For more details, see:
+    https://chromium.googlesource.com/chromium/src.git/+/HEAD/content/web_test/browser/test_info_extractor.h
+    """
+    # Marker sent by content_shell after blocks.
+    eof_marker = "#EOF" + linesep
+
+    def teardown(self):
+        # Close the queue properly to avoid broken pipe spam in the log.
+        self.browser.stdin_queue.close()
+        self.browser.stdin_queue.join_thread()
+
+    def send_command(self, command):
+        """Sends a single `command`, i.e. a URL to open, to content_shell.
+        """
+        self.browser.stdin_queue.put((command + linesep).encode("utf-8"))
+
+    def read_errors(self):
+        """Reads the entire content of the stderr queue as is available right now (no blocking).
+        """
+        result = ""
+
+        while not self.browser.stderr_queue.empty():
+            # There is no potential for race conditions here because this is the only place
+            # where we read from the stderr queue.
+            error = self.browser.stderr_queue.get()
+            result += error.decode("utf-8", "replace") + "\n"
+
+        return result
+
+    def read_line(self, deadline=None, encoding=None, errors="strict", raise_crash=True):
+        """Reads a single line from the stdout queue. The read must succeed before `deadline` or
+        a TimeoutError is raised. The line is returned as a bytestring or optionally with the
+        specified `encoding`. If `raise_crash` is set, a CrashError is raised if the line
+        happens to be a crash message.
+        """
+        current_time = time()
+
+        if deadline and current_time > deadline:
+            raise TimeoutError()
+
+        try:
+            line = self.browser.stdout_queue.get(True,
+                    deadline - current_time if deadline else None)
+
+            if raise_crash and line.startswith(b"#CRASHED"):
+                raise CrashError()
+        except Empty:
+            raise TimeoutError()
+
+        return line.decode(encoding, errors) if encoding else line
+
+    def is_alive(self):
+        """Checks if content_shell is still working by running a test on `about:blank` with
+        a short 2 second timeout.
+        """
+        try:
+            self.do_test("about:blank", 2)
+            return True
+        except (TimeoutError, CrashError):
+            return False
+
+    def do_test(self, url, timeout=None):
```

Comment:
So it seems unusual that this doesn't implement the `Protocol` API but defines a totally new interface that's specific to ContentShell. I see that the protocol is quite different from the usual WebDriver-esque protocols, but I think it would be better to conform to the existing API to the extent possible.

Reply:
I don't see how we could conform to the `Protocol` API basically at all. The only thing that the content_shell testing protocol supports right now is sending a single URL and getting text, image, or audio output. We can run the tests but we cannot implement any of the `ProtocolPart` classes. So the custom logic that is shared between `Executor` classes has to go somewhere. I see four options:
1. Put the logic in the `Protocol` class.
2. Create a new `ProtocolPart` and put the logic in there.
3. Put the logic in a custom `ExecutorBrowser` class.
4. Put the logic in a completely separate class / functions.

From what I can see in the codebase, [executorselenium.py](https://github.com/web-platform-tests/wpt/blob/0d6d9dc80bf23f94fab8137920b73228690b6de7/tools/wptrunner/wptrunner/executors/executorselenium.py#L267) does (essentially) option 1 by adding a `webdriver` attribute to its `Protocol` and using that for custom logic in some places. [executorservo.py](https://github.com/web-platform-tests/wpt/blob/0d6d9dc80bf23f94fab8137920b73228690b6de7/tools/wptrunner/wptrunner/executors/executorservo.py#L72) does option 4 and just uses an empty `ConnectionlessProtocol` (which is what we inherit from).

Reply:
I have some tentative suggestions:
1. Currently, some details of protocol mode are leaked into `ContentShellBrowser` (e.g., the fact that the protocol is line-based, and that we wait for a `#READY` message). Maybe we could do the following:
    * Reduce `ContentShellBrowser` to managing the subprocess and exposing stdio handles.
    * `ContentShellProtocol` becomes a regular `Protocol` and manages the stdio queues/threads, where `connect()` waits for the `#READY` message.
1. IMO, it's kind of strange that there's a `do_test(...)` method on this protocol (although I guess you could argue this protocol is explicitly designed for testing). There are really two interactions with content shell: running the test and checking for errors on stderr, which I'm leaning towards modeling as separate `ProtocolPart`s (i.e., option 2).

Need to think about this some more.

Reply:
I've done an attempt at this, let me know if this looks better.

Reply:
Also, just to clarify, the reason why ContentShellBrowser manages the stdio queues and threads is because of the architecture of wptrunner. ContentShellProtocol and ContentShellBrowser are running in different processes. The queues are not just there to provide non-blocking IO but to allow access to the IO from across processes in a clean way that works on all platforms. Sharing the handles themselves requires a bunch of platform-dependent voodoo AFAIK. Leaking the line-based nature of the protocol into the browser class is well worth it to not deal with that imo.